src/compiler/checker.ts

```diff
@@ -34550,6 +34550,9 @@ namespace ts {
             }
 
             const location = isBinaryExpression(condExpr) ? condExpr.right : condExpr;
+            if (isBinaryExpression(condExpr) && condExpr.operatorToken.kind === SyntaxKind.BarBarToken) {

```

Comment:
should this work on `&&` too? It seems like it but I haven't thought about it too deeply.
Edit: From the issue it sounds like we want to enable one operator at a time to make sure we don't introduce lots of false positives.

Reply:
I agree. I think enabling this for `&&` too could also catch many bugs like in 

```ts
declare function isFoo(): boolean;
if (isFoo && somethingElse) {}
```

The added difficulty will be handling values that might be `undefined`. In the case of 

```ts
declare const isFoo: (() => boolean) | undefined;
if (isFoo && somethingElse) {}
```

we would not want to throw an error. This is a similar complexity to what we have with negations (#43096). We don't have this problem with the `||` operator because if the value is `undefined` we would just move on to the right operand anyway.

An additional problem might be with incomplete type definitions of libraries where functions lack the `undefined` type. I'm not sure how big of a problem this might turn out to be and I don't think we should let this inform too much of what we do here but it is something to keep in mind.

Reply:
We do already check conjunctions but there is a related case where we do not check values that are referenced in the block. I opened #43211 to maybe do some experimentation on that front.